cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,549 @@
+
+
+def report_significant_avg_delta(key, orig_sample, new_sample):
+    """Return dict describing service boottime deltas when significant.
+
+    Compare orig_sample to new_sample average boottimes. If the percentage
+    delta between orig and new are significant, return a string classifying
+    whether the delta is IMPROVED or DEGRADED in comparison to the orig_sample.
+
+    :return: Empty dict when no significant delta present. Otherwise, a
+        dict containing significant delta characteristics:
+           name, orig_avg, new_avg,
+        DEGRADED boottimes for this comparison. Return empty string if no
+        significant difference exists.
+    """
+    avg_delta_time = new_sample["avg"] - orig_sample["avg"]
+    if orig_sample["avg"] > 0:
+        avg_delta_pct = new_sample["avg"] / orig_sample["avg"] * 100 - 100
+    else:
+        avg_delta_pct = 100
+    if abs(avg_delta_time) < MIN_AVG_DELTA:
+        if key not in MANDATORY_REPORT_KEYS:
+            # Always want to print the above keys
+            return {}
+    if abs(avg_delta_pct) < MIN_AVG_PERCENT_DELTA:
+        if key not in MANDATORY_REPORT_KEYS:
+            # Always want to print the above keys
+            return {}
+    if abs(avg_delta_pct) < MIN_AVG_PERCENT_DELTA and abs(avg_delta_time) < 2:
+        delta_type = "        "
+    elif avg_delta_pct > 0:
+        delta_type = "DEGRADED"
+    else:
+        delta_type = "IMPROVED"
+    return {
+        "orig_avg": orig_sample["avg"],
+        "orig_stdev": orig_sample["stdev"],
+        "new_avg": new_sample["avg"],
+        "new_stdev": new_sample["stdev"],
+        "avg_delta": new_sample["avg"] - orig_sample["avg"],
+        "name": key,
+        "delta_type": delta_type,
+    }
+
+
+def inspect_averages(orig_avg, new_avg):
+
+    perf_deltas = []
+
+    for key in orig_avg:
+        if "avg" in orig_avg[key] and orig_avg[key]["avg"] > MIN_AVG_DELTA:
+            perf_delta = report_significant_avg_delta(key, orig_avg[key], new_avg[key])
+            if perf_delta:
+                perf_deltas.append(PERF_LINE.format(**perf_delta))
+        else:
+            for service in set(orig_avg[key].keys()).union(new_avg.keys()):
+                if service in orig_avg[key] and service in new_avg[key]:
+                    perf_delta = report_significant_avg_delta(
+                        service, orig_avg[key][service], new_avg[key][service]
+                    )
+                    if perf_delta:
+                        perf_deltas.append(PERF_LINE.format(**perf_delta))
+    if perf_deltas:
+        perf_deltas.insert(0, PERF_HEADER)
+    for delta in perf_deltas:
+        print(delta)
+    if not perf_deltas:
+        print("------ No significant boot speed deltas seen -----")
+    log_avg_table("Control", orig_avg)
+    log_avg_table("Updated cloud-init", new_avg)
+
+
+def inspect_boot_time_deltas_for_deb(
```

Comment:
I think `main()` would make more sense as a function name - I wouldn't expect image build to be happening in a function with this main.